packages/chatbox/src/chatbox.ts

```diff
@@ -0,0 +1,493 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  each
+} from '@phosphor/algorithm';
+
+import {
+  DisposableSet
+} from '@phosphor/disposable';
+
+import {
+  Message
+} from '@phosphor/messaging';
+
+import {
+  Panel, PanelLayout, Widget
+} from '@phosphor/widgets';
+
+import {
+} from '@phosphor/widgets';
+
+import {
+  DocumentRegistry
+} from '@jupyterlab/docregistry';
+
+import {
+  CodeEditor
+} from '@jupyterlab/codeeditor';
+
+import {
+  BaseCellWidget,
+  MarkdownCellModel, MarkdownCellWidget
+} from '@jupyterlab/cells';
+
+import {
+  IObservableVector, ObservableVector
+} from '@jupyterlab/coreutils';
+
+import {
+  IRenderMime
+} from '@jupyterlab/rendermime';
+
+import {
+  ChatEntry
+} from './entry';
+
+
+/**
+ * The class name added to chatbox widgets.
+ */
+const CHATBOX_CLASS = 'jp-Chatbox';
+
+/**
+ * The class name of the active prompt
+ */
+const PROMPT_CLASS = 'jp-Chatbox-prompt';
+
+/**
+ * The class name of the panel that holds cell content.
+ */
+const CONTENT_CLASS = 'jp-Chatbox-content';
+
+/**
+ * The class name of the panel that holds prompts.
+ */
+const INPUT_CLASS = 'jp-Chatbox-input';
+
+
+/**
+ * A widget containing a Jupyter chatbox.
+ *
+ * #### Notes
+ * The Chatbox class is intended to be used within a ChatboxPanel
+ * instance. Under most circumstances, it is not instantiated by user code.
+ */
+export
+class Chatbox extends Widget {
+  /**
+   * Construct a chatbox widget.
+   */
+  constructor(options: Chatbox.IOptions) {
+    super();
+    this.addClass(CHATBOX_CLASS);
+
+    // Create the panels that hold the content and input.
+    let layout = this.layout = new PanelLayout();
+    this._content = new Panel();
+    this._input = new Panel();
+
+    this.contentFactory = options.contentFactory;
+    this.rendermime = options.rendermime;
+
+    // Add top-level CSS classes.
+    this._content.addClass(CONTENT_CLASS);
+    this._input.addClass(INPUT_CLASS);
+
+    // Insert the content and input panes into the widget.
+    layout.addWidget(this._content);
+    layout.addWidget(this._input);
+  }
+
+  /**
+   * The content factory used by the chatbox.
+   */
+  readonly contentFactory: Chatbox.IContentFactory;
+
+  /**
+   * The rendermime instance used by the chatbox.
+   */
+  readonly rendermime: IRenderMime;
+
+  /**
+   * Whether the chatbox has been disposed.
+   */
+  get isDisposed(): boolean {
+    return this._disposables === null;
+  }
+
+  /*
+   * The chatbox input prompt.
+   */
+  get prompt(): MarkdownCellWidget | null {
+    let inputLayout = (this._input.layout as PanelLayout);
+    return inputLayout.widgets[0] as MarkdownCellWidget || null;
+  }
+
+  /**
+   * The document model associated with the chatbox.
+   */
+  get model(): DocumentRegistry.IModel {
+    return this._model;
+  }
+  set model(model: DocumentRegistry.IModel) {
+    this._model = model;
+    let modelDB: any = (this._model as any).modelDB;
```

Comment:
Yeah, there are a few temporary stubs in this PR that I will take care of after the #2058 is resolved.